### DISCUSSION TITLE
Problem creating/running an agent

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#2285] [Q&A]
This morning I cloned **agent-starter** and followed the instructions to create a **.env** file.  I also created  and agent application, noted the CLINET_ID and CLIENT_SECRET and put these in my .env file.  Finally I added the client URL as a user to my test project.  

When I try to run the agent I get the following error.  I tried again several hours later and got the same result.

Are there anything else I need to do to get this to run?

Thanks.

**Output with Error**
**-------------------------------------------**

PS D:\iTwin\prototypes\phocaz-itwin-agent> npm start -- --latest

> my-agent@0.0.1 start D:\iTwin\prototypes\phocaz-itwin-agent
> node ./lib/main.js "--latest"

MY AGENT STARTED
Error   |imodelhub-client.iModelHub| 403 iModelHub.UserDoesNotHavePermission: User b87326cf-cae2-4b52-817b-fb8322d54499 does not have permission to context or iModel bbc2bbda-df82-45a9-805e-dcaebb055ff4.
iModelHub.UserDoesNotHavePermission: User b87326cf-cae2-4b52-817b-fb8322d54499 does not have permission to context or iModel bbc2bbda-df82-45a9-805e-dcaebb055ff4.
    at parse (D:\iTwin\prototypes\phocaz-itwin-agent\node_modules\@bentley\imodelhub-client\lib\imodelhub\Errors.js:82:27)
    at Object.request (D:\iTwin\prototypes\phocaz-itwin-agent\node_modules\@bentley\itwin-client\lib\Request.js:290:29)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  errorNumber: 102404,
  _getMetaData: undefined,
  description: 'cannot POST /imodelhub/sv1.1/Repositories/iModel--bbc2bbda-df82-45a9-805e-dcaebb055ff4/iModelScope/Briefcase/ (403)',
  _data: {
    errorMessage: 'User b87326cf-cae2-4b52-817b-fb8322d54499 does not have permission to context or iModel bbc2bbda-df82-45a9-805e-dcaebb055ff4. '
  },
  status: 403
} Unhandled exception thrown in my-agent
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! my-agent@0.0.1 start: `node ./lib/main.js "--latest"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the my-agent@0.0.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\johnm\AppData\Roaming\npm-cache\_logs\2021-09-17T19_54_36_170Z-debug.log
PS D:\iTwin\prototypes\phocaz-itwin-agent> npm start -- --latest

> my-agent@0.0.1 start D:\iTwin\prototypes\phocaz-itwin-agent
> node ./lib/main.js "--latest"

MY AGENT STARTED
Error   |imodelhub-client.iModelHub| 403 iModelHub.UserDoesNotHavePermission: User b87326cf-cae2-4b52-817b-fb8322d54499 does not have permission to context or iModel bbc2bbda-df82-45a9-805e-dcaebb055ff4.
iModelHub.UserDoesNotHavePermission: User b87326cf-cae2-4b52-817b-fb8322d54499 does not have permission to context or iModel bbc2bbda-df82-45a9-805e-dcaebb055ff4.
    at parse (D:\iTwin\prototypes\phocaz-itwin-agent\node_modules\@bentley\imodelhub-client\lib\imodelhub\Errors.js:82:27)
    at Object.request (D:\iTwin\prototypes\phocaz-itwin-agent\node_modules\@bentley\itwin-client\lib\Request.js:290:29)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  errorNumber: 102404,
  _getMetaData: undefined,
  description: 'cannot POST /imodelhub/sv1.1/Repositories/iModel--bbc2bbda-df82-45a9-805e-dcaebb055ff4/iModelScope/Briefcase/ (403)',
  _data: {
    errorId: 'iModelHub.UserDoesNotHavePermission',
    errorMessage: 'User b87326cf-cae2-4b52-817b-fb8322d54499 does not have permission to context or iModel bbc2bbda-df82-45a9-805e-dcaebb055ff4. '
  },
  status: 403
} Unhandled exception thrown in my-agent
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! my-agent@0.0.1 start: `node ./lib/main.js "--latest"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the my-agent@0.0.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\johnm\AppData\Roaming\npm-cache\_logs\2021-09-17T19_55_35_452Z-debug.log

**Comment by JohnMajerle** [comment]
I tried running my agent services this morning and got a different error:

```
node ./lib/main.js "--latest"

MY AGENT STARTED
Error   |imodelhub-client.iModelHub| 501 iModelHub.FeatureIsDisabled: Feature is disabled.
CLIP-TEST-01
MY AGENT FINISHED
```

**Comment by aruniverse** [comment]
Hey John, just want to confirm a few things:
- What's the auth issuer you're using? Is it `ims.bentley.com`? Take a look at https://github.com/imodeljs/imodeljs/discussions/2158, specifically the section for client-credential (agent) workflow at the bottom
- Did you add your service user, should look like `{your-client-id}@apps.imsoidc.bentley.com`, as a member of the project/imodel you are trying to access?
- Do you know what "feature" you are trying to use?

**Comment by JohnMajerle** [comment]
Hello Arun,
I cloned agent-starter and didn't make any code changes.  All I did was add the required .env file required by the instructions.  In that file I have:
imjs_itwin_platform_authority=https://ims.bentley.com
As mentioned in that link you sent, **_MyAgent.ts_** uses the scope **_itwinjs_**.
And I added my agent email as a user to this project:
service-8NYG1YdSOZazU2JKYcCOpURnH@apps.imsoidc.bentley.com

I verified that this agent has the  **_itwinjs_** scope included:
![image](https://user-images.githubusercontent.com/30668739/134011227-23b6f59a-81a6-4967-a340-d9beee6e36a5.png)

**Reply by aruniverse** [reply]
I just cloned the agent starter myself and it was working fine. Make sure the 
> All I did was add the required .env file required by the instructions. In that file I have:
imjs_itwin_platform_authority=ims.bentley.com

I think you're missing the `https://` which might be messing up the requests. Could you try setting it like so
`imjs_itwin_platform_authority="https://ims.bentley.com"`

It did take about 15-20 minutes for my agent to work properly because it took some time for the project member configuration(adding my service client to a project) to propagate through our system, might be why you saw those errors Friday

**Comment by JohnMajerle** [comment]
Hello Arun,

If you read my previous comment carefully above I stated:

imjs_itwin_platform_authority=https://ims.bentley.com

I see that in your reply you added quotes:

imjs_itwin_platform_authority="https://ims.bentley.com"

I tried that but got the same error:

MY AGENT STARTED
Error   |imodelhub-client.iModelHub| 501 iModelHub.FeatureIsDisabled: Feature is disabled.
CLIP-TEST-01
MY AGENT FINISHED

Are agent services considered an early access feature that someone at Bentley needs to enable for our company and projects?  If so perhaps that is why it worked for you but not for me?

**Reply by aruniverse** [reply]
> I see that in your reply you added quotes:

I also added `https://` as a prefix. Could you try that as well? Agent shouldn't be an early access feature, but I'll ask around with someone more familiar.

**Comment by JohnMajerle** [comment]
Hello Arun,

We seam to have a communication problem.  I do have https:// as a prefix.  Perhaps  your browser is not displaying correctly?

Below I have pasted a screen shot of my .env file with the client secret scribbled out:

![image](https://user-images.githubusercontent.com/30668739/134054054-d27998c3-73e8-41be-8598-bf41599459fd.png)

**Reply by aruniverse** [reply]
My apologies John, looks like on my end the https:// prefix was stripped. 
Continuing to investigate what could be going wrong...

**Comment by JohnMajerle** [comment]
Hello Arun,

Are agent services considered an early access feature that someone at Bentley needs to enable for our company and projects? If so perhaps that is why it worked for you but not for me?

**Reply by aruniverse** [reply]
Hey John, sorry for the delayed response, was out sick last week when you asked. 

Agents are not an early access feature, and they should work for external users such as yourself. Are you still running into issues?

**Reply by JohnMajerle** [reply]
Hello Arun,

Yes.  I am getting the following error message displayed when I run the agent.  What would cause iModelHub to return FeatureIsDisabled?

--John M.

PS D:\iTwin\prototypes\phocaz-itwin-agent> npm start -- --latest

> my-agent@0.0.1 start D:\iTwin\prototypes\phocaz-itwin-agent
> node ./lib/main.js "--latest"

MY AGENT STARTED
Error   |imodelhub-client.iModelHub| 501 iModelHub.FeatureIsDisabled: Feature is disabled.
CLIP-TEST-01
MY AGENT FINISHED

**Comment by nn15296** [comment]
Hi @aruniverse 

I'm getting a similar issue. It occurs when I try to download the briefcase. I've obtained the full error shown below:

```
iModelHub.FeatureIsDisabled: Feature is disabled.
    at parse (C:\Users\NOR92037\Documents\AIDB2.0\node_modules\@bentley\imodelhub-client\lib\imodelhub\Errors.js:82:27)
    at Object.request (C:\Users\NOR92037\Documents\AIDB2.0\node_modules\@bentley\itwin-client\lib\Request.js:290:29)
    at processTicksAndRejections (internal/process/task_queues.js:94:5) {
  errorNumber: 102401,
  _getMetaData: undefined,
  name: 'iModelHub.FeatureIsDisabled',
  description: 'cannot GET /imodelhub/sv1.1/Repositories/iModel--451bac78-439c-4399-a725-a9f491346ac3/iModelScope/CheckpointV2/?$filter=PrecedingCheckpointV2-backward-ChangeSet.Id+eq+%27%27&$select=*,HasContainer-forward-ContainerAccessKey.* (501)',
  message: 'Feature is disabled.',
  _data: {
    errorId: 'iModelHub.FeatureIsDisabled',
    errorMessage: 'Feature is disabled.'
  },
  status: 501
}
```

I can confirm I have the following line in my ".env" file:

```
imjs_itwin_platform_authority=https://ims.bentley.com
```

The error is triggered when I try to download the briefcase. The code I use for this is below:

```
const progressBar = new SingleBar({ format: ` [${getTimeNow()}]  Downloading {bar} {percentage}% | ETA: {eta}s` }, Presets.shades_classic);
let barStarted: boolean = false;
const downloadedBriefcaseProps = await BriefcaseManager.downloadBriefcase(ctx, {
  contextId: this.config.CONTEXT_ID, iModelId: imodelId, asOf: version.toJSON(), onProgress: (loaded: number, total: number): number => {
    if (!barStarted) {
      progressBar.start(total, 0);
      barStarted = true;
    }
    progressBar.update(loaded);
    // Return non-zero to abort download
    return 0;
  },
});
```

**Reply by nn15296** [reply]
Hi @ramanujam-raman, just tagging you in case Arun is away.

**Reply by aruniverse** [reply]
Hi @nn15296 I've asked some devs who are more familiar with agents than I am. I was told the above error you're seeing is a non fatal error, and should fallback correctly and actually download the briefcase. We're having some devs looking into this, and they'll update this thread once they know more.

cc @JohnMajerle
